keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,291 @@
+
+The automated migration should work for Kubernetes built-in resource types,
+custom resources, and aggregated resources.
+
+Here are two alternatives and their requirements.
+
+#### Exposing the configured storage versions via API
+
+This is what the [Active Management][] design proposes. It requires:
+* exposing the configured [storage versions][] of Kubernetes built-in resource types and
+aggregated resources. 
+
+Note that custom resources already exposes this information. 
+
+The migration then is driven by the difference between the configured storage
+versions and the stored versions recorded in the "progress".
+
+[Active Management]:https://docs.google.com/document/d/1eoS1K40HLMl4zUyw5pnC05dEF3mzFLp5TPEEt4PFvsM
+[storage versions]:https://github.com/kubernetes/kubernetes/blob/db9545e69e630e5bc8c3c2dab14f9339dd3b871c/pkg/kubeapiserver/options/storage_versions.go#L98
+
+#### Using the apiserver version to trigger migration
+
+If the Kubernetes deprecation [policy][] is followed by every apiserver, and
+every cluster uses the default [storage version configs][], then we can use the
+apiserver version to drive the migration. More specifically,
+* kube-apiserver removes the `--storagve-version` flag, so that every deployment
+  of kube-apiserver uses the default storage versions.
+* aggregated apiservers follow semantic versioning, and expose the version via
+  API.
+* aggregated apiservers must not change storage version configurations at patch
+  versions.
+
+The migration for custom resources is still triggered by the different between
+its [storage version][] and the recorded [stored versions][].
+
+[storage version configs]:https://github.com/kubernetes/kubernetes/blob/db9545e69e630e5bc8c3c2dab14f9339dd3b871c/pkg/kubeapiserver/options/storage_versions.go#L98
+[storage version]:https://github.com/kubernetes/kubernetes/blob/c1f7df2b0efe89fb9d26db2f2837bac11d57064e/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L66
+[stored version]:https://github.com/kubernetes/kubernetes/blob/c1f7df2b0efe89fb9d26db2f2837bac11d57064e/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L177
+[policy]:https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+In both alternatives, HA apiservers present unique challenges. The migration
+must start after all HA apiservers have agreed on the storage versions or the
+apiserver versions. The agreement protocol requires its own KEP.
+
+We will revisit this section in 2018 Q4. 
+
+### Risks and Mitigations
+
+The migration process does not change the objects, so it will not pollute
+existing data.
+
+If the rate limiting is not tuned well, the migration can overload the
+apiserver. Users can delete the migration controller and the migration
+jobs to mitigate.
+
+Before upgrading or downgrading the cluster, the cluster administrator MUST
+check the "progress" object, and make sure every resource type is marked as
```

Comment:
We can do slightly better than this. It can state separately whether roll-back and roll-forward is safe.

E.g., if you roll-forward, and then roll-back before the tool is finished, that is actually fine. You just can't roll-forward twice in a row without waiting, nor can you roll back twice in a row.

The tool has to track state to understand whether the operations in progress are rolling forward or backwards to support this.

I think it might actually be quite important, especially in large clusters, to be able to do an immediate rollback without waiting for a data migration.

Reply:
> can we deprecate --storage-versions? It's defined in pkg/kubeapiserver. Kuberentes is following the 4-step deprecation policy, I don't see in what case the --storage-versions flag is useful

I agree with this. It seems like flexibility we don't need, that actually harms our ability to reason about what is safe to do as we maintain kube-apiserver in future releases.

Reply:
@lavalamp @deads2k what do you think about removing ` --storage-versions`? I opened https://github.com/kubernetes/kubernetes/pull/67678 to remove the flag, but it's more complicated than I expected, can you let me know early if you think it's a bad idea?

Reply:
removing `--storage-versions` means everyone has to use the default. How would you e.g. do a safe downgrade if someone skipped a step in the version upgrade process?

I think most people should use the default.

We can't remove the flag without a deprecation process unless we exhaustively demonstrate that no one is using it.

Reply:
> How would you e.g. do a safe downgrade if someone skipped a step in the version upgrade process?

Not sure if I follow, could you be more specific?

> We can't remove the flag without a deprecation process unless we exhaustively demonstrate that no one is using it.

How about keeping a noop `--storage-versions` flag for one more release?

side note: if users are using this flag, they need to manually inspect this flag every time they upgrade anyway, because the API version in the flag value can be deprecated.

Reply:
One more release is not our deprecation policy. Without looking, I think we need at least two, maybe more.

If someone introduces a v2 and also accidentally sets it as the storage version in the same release, it is impossible to downgrade safely without use of the storage version flag.

Reply:
> If someone introduces a v2 and also accidentally sets it as the storage version in the same release, it is impossible to downgrade safely without use of the storage version flag.

We should write tests to prevent this kind of bugs. However, as we don't expose the storage versions, I don't know how to write such tests.

The deprecation [policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli) says 6 months.

Reply:
Just realized that after we remove the `--storage-versions`, a group's preferred version showed in the discovery API equals the storage version, then we don't need to argue if we should expose the storage versions.. Is there any other reason why storage version might diverge from preferred version?